pkg/runtest/run.go

```diff
@@ -59,94 +57,19 @@ type Context struct {
 	Tests        string // prefix to match test file names
 
 	executor queue.PlainQueue
+	requests []*runRequest
+	buildSem chan bool
 }
 
 func (ctx *Context) log(msg string, args ...interface{}) {
 	ctx.LogFunc(fmt.Sprintf(msg, args...))
 }
 
 func (ctx *Context) Run() error {
-	if ctx.Retries%2 == 0 {
-		ctx.Retries++
-	}
-	progs := make(chan *runRequest, 1000)
-	var eg errgroup.Group
-	eg.Go(func() error {
-		defer close(progs)
-		return ctx.generatePrograms(progs)
-	})
-	done := make(chan *runRequest)
-	eg.Go(func() error {
-		return ctx.processResults(done)
-	})
-
-	var wg sync.WaitGroup
-	for req := range progs {
-		req := req
-		if req.broken != "" || req.skip != "" {
-			done <- req
-			continue
-		}
-		var retry queue.DoneCallback
-		retry = func(_ *queue.Request, res *queue.Result) bool {
-			// The tests depend on timings and may be flaky, esp on overloaded/slow machines.
-			// We don't want to fix this by significantly bumping all timeouts,
-			// because if a program fails all the time with the default timeouts,
-			// it will also fail during fuzzing. And we want to ensure that it's not the case.
-			// So what we want is to tolerate episodic failures with the default timeouts.
-			// To achieve this we run each test several times and ensure that it passes
-			// in 50+% of cases (i.e. 1/1, 2/3, 3/5, 4/7, etc).
-			// In the best case this allows to get off with just 1 test run.
-
-			if res.Err != nil {
-				req.err = res.Err
-				return true
-			}
-			req.result = res
-			err := checkResult(req)
-			if err == nil {
-				req.ok++
-			} else {
-				req.failed++
-				req.err = err
-			}
-			if req.ok > req.failed {
-				// There are more successful than failed runs.
-				req.err = nil
-				return true
-			}
-			// We need at least `failed - ok + 1` more runs <=> `failed + ok + need` in total,
-			// which simplifies to `failed * 2 + 1`.
-			if req.failed*2+1 <= ctx.Retries {
-				// We can still retry the execution.
-				req.OnDone(retry)
-				ctx.executor.Submit(req.Request)
-				return false
-			}
-			// Give up and fail on this request.
-			return true
-		}
-		req.Request.OnDone(retry)
-		ctx.executor.Submit(req.Request)
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			req.Request.Wait(context.Background())
-			done <- req
-		}()
-	}
-	wg.Wait()
-	close(done)
-	return eg.Wait()
-}
-
-func (ctx *Context) Next() *queue.Request {
-	return ctx.executor.Next()
-}
-
-func (ctx *Context) processResults(requests chan *runRequest) error {
+	ctx.buildSem = make(chan bool, runtime.GOMAXPROCS(0))
+	ctx.generatePrograms()
 	var ok, fail, broken, skip int
-	for req := range requests {
```

Comment:
Do you mean w/o having to wait for _all_ results?

The current code also prints the first test result as soon as it's ready (it's also submitted to the queue first).

This does change behavior in presence of retries, however. But in exchange to much faster total execution time.

We could use a prio queue to prioritize tries of earlier tests. Then it will the best of both worlds.

Reply:
Added request priority.